File: tests/ut/cpp/types/test_buffer.cpp
/*
 * Copyright (c) PyPTO Contributors.
 * This program is free software, you can redistribute it and/or modify it under the terms and conditions of
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 * -----------------------------------------------------------------------------------------------------------
 */
// Wire-ABI tests for Buffer / Tensor (buffer.h); the contract they pin is described
// in docs/buffer-abi.md. Byte layout is pinned by static_assert in the header; these tests
// pin the sizes, enum values, and the shared receive-side gate from the outside.

#include <cstddef>
#include <cstdint>
#include <cstring>
#include <stdexcept>
#include <string>
#include <vector>

#include <gtest/gtest.h>

#include "buffer.h"
#include <random>
#include <vector>

#include "task_args.h"

namespace {

CanonicalIdentity make_identity() {
    CanonicalIdentity id{};
    for (uint32_t i = 0; i < OWNER_INSTANCE_ID_BYTES; ++i)
        id.owner_instance_id[i] = static_cast<uint8_t>(0xA0 + i);
    id.buffer_id = 0x0102030405060708ULL;
    id.generation = 7;
    return id;
}

Tensor make_tensor() {
    Tensor r{};
    r.buffer.magic = BUFFER_DESCRIPTOR_MAGIC;
    r.buffer.backend_kind = static_cast<uint8_t>(BackendKind::POSIX_SHM);
    r.buffer.identity = make_identity();
    r.buffer.nbytes = 8192;  // must cover byte_offset + the strided extent below
    r.byte_offset = 4096;
    r.ndims = 3;
    r.shapes[0] = 2;
    r.shapes[1] = 4;
    r.shapes[2] = 8;
    r.strides[0] = 32;
    r.strides[1] = 8;
    r.strides[2] = 1;
    r.dtype = DataType::FLOAT16;
    return r;
}

// --- Layout / value contracts (frozen ABI) -----------------------------------------------------

TEST(BufferAbi, StructSizesAreFrozen) {
    EXPECT_EQ(sizeof(CanonicalIdentity), 32u);
    EXPECT_EQ(sizeof(Tensor), 144u);
    EXPECT_EQ(sizeof(BufferDescriptor), 88u);
}

TEST(BufferAbi, ConstantsAreFrozen) {
    EXPECT_EQ(BUFFER_DESCRIPTOR_MAGIC, 0x5342);
    EXPECT_EQ(OWNER_INSTANCE_ID_BYTES, 8u);
    EXPECT_EQ(DESC_MAX_BYTES, 32u);
}

TEST(BufferAbi, EnumValuesAreFrozen) {
    EXPECT_EQ(static_cast<uint8_t>(AddressSpace::HOST), 0);
    EXPECT_EQ(static_cast<uint8_t>(AddressSpace::DEVICE), 1);
    EXPECT_EQ(static_cast<uint8_t>(AccessMode::READ), 0);
    EXPECT_EQ(static_cast<uint8_t>(AccessMode::WRITE), 1);
    EXPECT_EQ(static_cast<uint8_t>(AccessMode::READWRITE), 2);
    EXPECT_EQ(static_cast<uint8_t>(BackendKind::FORK_SHM), 0);
    EXPECT_EQ(static_cast<uint8_t>(BackendKind::POSIX_SHM), 1);
    EXPECT_EQ(static_cast<uint8_t>(BackendKind::VMM_WINDOW), 2);
    EXPECT_EQ(static_cast<uint8_t>(BackendKind::REMOTE_SIDECAR), 3);
    EXPECT_EQ(static_cast<uint8_t>(BackendKind::DEVICE_MALLOC), 4);
    EXPECT_EQ(static_cast<uint8_t>(BackendKind::FORK_COW), 5);
}

// --- memcpy round trip -------------------------------------------------------------------------

TEST(BufferAbi, TensorSurvivesByteRoundTrip) {
    Tensor src = make_tensor();
    uint8_t bytes[sizeof(Tensor)];
    std::memcpy(bytes, &src, sizeof(Tensor));
    Tensor dst{};
    std::memcpy(&dst, bytes, sizeof(Tensor));
    EXPECT_EQ(std::memcmp(&src, &dst, sizeof(Tensor)), 0);
    EXPECT_EQ(dst.byte_offset, 4096u);
    EXPECT_EQ(dst.dtype, DataType::FLOAT16);
    EXPECT_EQ(dst.strides[0], 32u);
    EXPECT_EQ(dst.buffer.identity, src.buffer.identity);
}

TEST(BufferAbi, DescriptorSurvivesByteRoundTrip) {
    BufferDescriptor src{};
    src.magic = BUFFER_DESCRIPTOR_MAGIC;
    src.address_space = static_cast<uint8_t>(AddressSpace::DEVICE);
    src.access = static_cast<uint8_t>(AccessMode::READWRITE);
    src.backend_kind = static_cast<uint8_t>(BackendKind::POSIX_SHM);
    src.owner_worker_path_id = 3;
    src.identity = make_identity();
    src.nbytes = 1 << 20;
    const char *body = "psm_deadbeef";
    src.body_len = static_cast<uint16_t>(std::strlen(body));
    std::memcpy(src.body, body, src.body_len);

    uint8_t bytes[sizeof(BufferDescriptor)];
    std::memcpy(bytes, &src, sizeof(BufferDescriptor));
    BufferDescriptor dst{};
    std::memcpy(&dst, bytes, sizeof(BufferDescriptor));
    EXPECT_EQ(std::memcmp(&src, &dst, sizeof(BufferDescriptor)), 0);
    EXPECT_EQ(dst.magic, BUFFER_DESCRIPTOR_MAGIC);
    EXPECT_EQ(dst.owner_worker_path_id, 3u);
    EXPECT_EQ(dst.identity, src.identity);
    EXPECT_EQ(std::string(dst.body, dst.body_len), "psm_deadbeef");
}

// --- canonical identity: the import-registry key -----------------------------------------------

TEST(BufferAbi, IdentityDistinguishesGenerationAndIncarnation) {
    CanonicalIdentity a = make_identity();
    CanonicalIdentity b = make_identity();
    EXPECT_EQ(a, b);

    b.generation = a.generation + 1;  // buffer_id reuse across generations (ABA)
    EXPECT_NE(a, b);

    CanonicalIdentity c = make_identity();
    c.owner_instance_id[0] ^= 0xFF;  // different owner incarnation nonce
    EXPECT_NE(a, c);

    CanonicalIdentity d = make_identity();
    d.buffer_id ^= 0xFFULL;
    EXPECT_NE(a, d);
}

// Padding is wire-visible but semantically absent: two decodes of one backing must key identically,
// or the same buffer lands in two import-registry buckets and its dependencies split.
TEST(BufferAbi, IdentityPaddingIsExcludedFromKeyAndHash) {
    CanonicalIdentityHash h;
    CanonicalIdentity clean = make_identity();
    CanonicalIdentity dirty = make_identity();
    std::memset(dirty._pad, 0xA5, sizeof(dirty._pad));
    EXPECT_EQ(clean, dirty);
    EXPECT_EQ(h(clean), h(dirty));
}

TEST(BufferAbi, IdentityHashMatchesEquality) {
    CanonicalIdentityHash h;
    CanonicalIdentity a = make_identity();
    CanonicalIdentity b = make_identity();
    EXPECT_EQ(h(a), h(b));

    b.generation = a.generation + 1;
    EXPECT_NE(h(a), h(b));  // good hash separates the ABA case (not a strict requirement)
}

// --- validate_tensor: the shared receive-side gate ------------------------------------------

TEST(BufferAbi, ForkCowGrantsReadOnly) {
    Tensor r = make_tensor();
    r.buffer.backend_kind = static_cast<uint8_t>(BackendKind::FORK_COW);
    r.buffer.access = static_cast<uint8_t>(AccessMode::READ);
    EXPECT_NO_THROW(validate_tensor(r));

    // A write grant over copy-on-write would land in a private copy the owner never sees, so it is
    // refused rather than silently dropped.
    for (auto bad : {AccessMode::WRITE, AccessMode::READWRITE}) {
        r.buffer.access = static_cast<uint8_t>(bad);
        EXPECT_THROW(validate_tensor(r), std::invalid_argument);
    }

    // The same grants stay legal over MAP_SHARED, where a child's write does reach the owner.
    r.buffer.backend_kind = static_cast<uint8_t>(BackendKind::FORK_SHM);
    r.buffer.access = static_cast<uint8_t>(AccessMode::READWRITE);
    EXPECT_NO_THROW(validate_tensor(r));
}

// --- Malformed wire bytes: what only a decoder can produce ------------------------------------
//
// A receiver memcpy's an element out of shared memory a peer wrote and then validates it, so every
// field arrives as whatever bytes were there. These cases are unreachable through construction (the
// constructor validates first) and unreachable from Python (no binding turns bytes into a Tensor),
// which is why they live here: this is the only place the malformed state can be built at all.

// Overwrite `len` bytes at `offset` of a packed Tensor, then decode it the way a receiver does.
Tensor decode_with_corruption(size_t offset, const uint8_t *value, size_t len) {
    uint8_t bytes[sizeof(Tensor)];
    Tensor src = make_tensor();
    std::memcpy(bytes, &src, sizeof(Tensor));
    if (len > 0) std::memcpy(bytes + offset, value, len);
    Tensor out;
    std::memcpy(&out, bytes, sizeof(Tensor));
    return out;
}

TEST(BufferAbi, DecodeRejectsEveryCorruptedField) {
    // The untouched element is accepted, so each rejection below is attributable to the one field.
    EXPECT_NO_THROW(validate_tensor(decode_with_corruption(0, nullptr, 0)));  // no-op corruption

    struct Case {
        const char *what;
        size_t offset;
        std::vector<uint8_t> value;
    };
    // Offsets are the frozen ones the static_asserts above pin.
    const Case cases[] = {
        {"descriptor magic", offsetof(Tensor, buffer) + offsetof(BufferDescriptor, magic), {0x00, 0x00}},
        {"address_space out of range", offsetof(BufferDescriptor, address_space), {0x07}},
        {"access out of range", offsetof(BufferDescriptor, access), {0x09}},
        {"backend_kind out of range", offsetof(BufferDescriptor, backend_kind), {0x63}},
        {"body_len past the array", offsetof(BufferDescriptor, body_len), {0xFF, 0xFF}},
        {"generation 0 is reserved",
         offsetof(BufferDescriptor, identity) + offsetof(CanonicalIdentity, generation),
         {0, 0, 0, 0}},
        {"ndims out of range", offsetof(Tensor, ndims), {0x63, 0, 0, 0}},
        {"unknown dtype", offsetof(Tensor, dtype), {0x7F}},
        {"stride must be > 0", offsetof(Tensor, strides), {0, 0, 0, 0}},
        {"byte_offset is not a multiple", offsetof(Tensor, byte_offset), {0x03, 0, 0, 0, 0, 0, 0, 0}},
    };
    for (const auto &c : cases) {
        SCOPED_TRACE(c.what);
        EXPECT_THROW(
            validate_tensor(decode_with_corruption(c.offset, c.value.data(), c.value.size())), std::invalid_argument
        );
    }
}

TEST(BufferAbi, DecodeRejectsAViewPastItsBacking) {
    Tensor r = make_tensor();
    r.buffer.nbytes = r.byte_offset + tensor_extent_bytes(r);
    EXPECT_NO_THROW(validate_tensor(r));  // exact fit

    r.buffer.nbytes -= 1;
    EXPECT_THROW(validate_tensor(r), std::invalid_argument);

    r = make_tensor();
    r.buffer.nbytes = r.byte_offset + tensor_extent_bytes(r);
    r.byte_offset += get_element_size(r.dtype);  // shifting the origin pushes the tail out
    EXPECT_THROW(validate_tensor(r), std::invalid_argument);
}

TEST(BufferAbi, ValidateNeverWalksOffArbitraryBytes) {
    // A peer can write anything of the right length. None of it may be accepted without passing
    // every field check, and none of it may read past the struct — the fixed-length identity and
    // the bounded body_len/ndims are what make that structural rather than a matter of luck.
    std::mt19937 rng(0x5342);  // fixed seed: a failure here must reproduce
    for (int i = 0; i < 4096; ++i) {
        uint8_t bytes[sizeof(Tensor)];
        for (auto &b : bytes)
            b = static_cast<uint8_t>(rng());
        Tensor r;
        std::memcpy(&r, bytes, sizeof(Tensor));
        try {
            validate_tensor(r);
        } catch (const std::invalid_argument &) {
            continue;  // the expected outcome
        }
        // Surviving is legal only if every invariant genuinely holds.
        EXPECT_EQ(r.buffer.magic, BUFFER_DESCRIPTOR_MAGIC);
        EXPECT_NE(r.buffer.identity.generation, 0u);
        EXPECT_LE(r.buffer.body_len, DESC_MAX_BYTES);
        EXPECT_GT(r.ndims, 0u);
        EXPECT_LE(r.ndims, static_cast<uint32_t>(MAX_TENSOR_DIMS));
    }

    for (uint8_t fill : {uint8_t{0x00}, uint8_t{0xFF}}) {
        uint8_t bytes[sizeof(Tensor)];
        std::memset(bytes, fill, sizeof(bytes));
        Tensor r;
        std::memcpy(&r, bytes, sizeof(Tensor));
        EXPECT_THROW(validate_tensor(r), std::invalid_argument);
    }
}

TEST(BufferAbi, DecodedIdentityIgnoresWirePadding) {
    // Two decodes of one backing must key identically however the padding arrived, or an import
    // registry splits one buffer across two entries.
    uint8_t clean[sizeof(CanonicalIdentity)];
    CanonicalIdentity src = make_identity();
    std::memcpy(clean, &src, sizeof(CanonicalIdentity));

    uint8_t dirty[sizeof(CanonicalIdentity)];
    std::memcpy(dirty, clean, sizeof(CanonicalIdentity));
    std::memset(dirty + offsetof(CanonicalIdentity, _pad), 0xA5, sizeof(CanonicalIdentity::_pad));

    CanonicalIdentity a;
    CanonicalIdentity b;
    std::memcpy(&a, clean, sizeof(a));
    std::memcpy(&b, dirty, sizeof(b));
    EXPECT_TRUE(a == b);
    EXPECT_EQ(CanonicalIdentityHash{}(a), CanonicalIdentityHash{}(b));
}

}  // namespace

File: tests/ut/py/test_buffer.py
# Copyright (c) PyPTO Contributors.
# This program is free software, you can redistribute it and/or modify it under the terms and conditions of
# CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.
# -----------------------------------------------------------------------------------------------------------
"""Unit tests for simpler.buffer: identity/descriptor pack-unpack + create/import round trip.

The three wire types are the C++ structs of buffer.h bound directly, so what is pinned here is the
Python-visible contract over them — construction rejects what `validate_buffer_descriptor` rejects,
a round trip through `pack`/`unpack` is the identity, and equality and hashing ignore wire padding.
Imports come from `simpler.buffer` because that is where a caller reaches them, alongside the
registry and the Buffer constructors that are genuinely defined there.
"""

import pytest
from _task_interface import OWNER_INSTANCE_ID_BYTES, DataType
from simpler.buffer import (
    AccessMode,
    AddressSpace,
    BackendKind,
    BufferDescriptor,
    CanonicalIdentity,
    ImportRegistry,
    Tensor,
    create_host_shared_buffer,
    intern_worker_path,
    mint_owner_instance_id,
    re_export,
    wrap_device_malloc,
)
from simpler.task_interface import ChipTensor

_OID = bytes(range(0xA0, 0xA0 + OWNER_INSTANCE_ID_BYTES))


def test_wire_tensor_and_device_pod_are_distinct_types():
    # The whole point of the two names: one address-free argument, one GM-address-bearing POD. An
    # alias here would give `Tensor` a second meaning and every later cutover would have to keep it.
    assert ChipTensor is not Tensor


def test_wire_tensor_stays_off_the_public_submit_surface():
    # `TaskArgs.add_tensor` still takes a ChipTensor, so re-exporting `Tensor` from
    # simpler.task_interface would advertise a public type that its own submit call rejects. The two
    # facts move together: the wire cutover that makes add_tensor accept a Tensor is what earns it a
    # place on that module, and this test fails then to say so.
    import simpler.task_interface as ti  # noqa: PLC0415

    assert not hasattr(ti, "Tensor")

    h = create_host_shared_buffer(64, mint_owner_instance_id(), buffer_id=1)
    try:
        args = ti.TaskArgs()
        with pytest.raises(TypeError):
            args.add_tensor(h.tensor(shapes=(16,), dtype=DataType.FLOAT32))
    finally:
        h.close()


def _identity(oid=_OID, buffer_id=7, generation=2):
    return CanonicalIdentity(oid, buffer_id, generation)


def test_identity_rejects_bad_oid_width():
    with pytest.raises(ValueError):
        CanonicalIdentity(b"\x00" * (OWNER_INSTANCE_ID_BYTES + 1), 1, 1)


def test_identity_distinguishes_generation_and_incarnation():
    a = _identity()
    assert a != _identity(generation=a.generation + 1)  # ABA
    assert a != _identity(oid=bytes(range(1, 1 + OWNER_INSTANCE_ID_BYTES)))  # different incarnation
    assert a != _identity(buffer_id=a.buffer_id + 1)


def _descriptor_with_path_id(path_id: int) -> BufferDescriptor:
    return BufferDescriptor(
        identity=_identity(),
        address_space=AddressSpace.HOST,
        access=AccessMode.READWRITE,
        backend_kind=BackendKind.POSIX_SHM,
        nbytes=8,
        owner_worker_path_id=path_id,
    )


def test_worker_path_is_diagnostic_and_survives_an_unknown_id():
    h = _descriptor_with_path_id(intern_worker_path("L4/L3[2]"))
    assert h.owner_worker_path == "L4/L3[2]"
    # An id minted in another process has no local text, and that is not an error.
    assert _descriptor_with_path_id(99_999).owner_worker_path == "<path#99999>"
    # The path takes no part in identity: two descriptors differing only by path are one backing.
    assert _descriptor_with_path_id(0).identity == h.identity


def test_descriptor_rejects_oversized_body():
    with pytest.raises(ValueError, match="body"):
        BufferDescriptor(
            identity=_identity(),
            address_space=AddressSpace.HOST,
            access=AccessMode.READWRITE,
            backend_kind=BackendKind.POSIX_SHM,
            nbytes=8,
            body=b"x" * 200,  # > DESC_MAX_BYTES
        )


def test_create_export_import_resolve_zero_copy():
    oid = mint_owner_instance_id()
    buffer = create_host_shared_buffer(nbytes=256, owner_instance_id=oid, buffer_id=1, owner_worker_path="L4")
    reg = ImportRegistry()
    try:
        assert buffer.backend_kind == BackendKind.POSIX_SHM
        imported = reg.materialize(buffer.to_descriptor())
        assert reg.resolve(buffer.identity).base == imported.base
        assert reg.materialize(buffer.to_descriptor()).base == imported.base  # map-once: same mapping
        assert imported.nbytes == 256
        owner_shm = buffer.shm
        consumer_shm = imported.shm
        assert owner_shm is not None
        assert consumer_shm is not None
        owner_buf = owner_shm.buf
        consumer_buf = consumer_shm.buf
        assert owner_buf is not None
        assert consumer_buf is not None
        owner_buf[:4] = b"\xde\xad\xbe\xef"
        assert bytes(consumer_buf[:4]) == b"\xde\xad\xbe\xef"
    finally:
        reg.close()
        buffer.close()


def test_resolve_unregistered_raises():
    reg = ImportRegistry()
    with pytest.raises(KeyError):
        reg.resolve(_identity())


def test_tensor_full_view_is_contiguous():
    oid = mint_owner_instance_id()
    h = create_host_shared_buffer(nbytes=1024, owner_instance_id=oid, buffer_id=1)
    try:
        # buffer.tensor(shape, dtype) is a contiguous full view: row-major strides, zero offset.
        v = h.tensor(shapes=(4, 8), dtype=DataType.FLOAT32)
        assert v.shapes == (4, 8)
        assert v.strides == (8, 1)
        assert v.ndims == 2
        assert v.byte_offset == 0
        # An explicit stride is carried verbatim; a singleton dim is never normalized away.
        strided = h.tensor(shapes=(4, 1), dtype=DataType.FLOAT32, strides=(8, 3))
        assert strided.strides == (8, 3)
    finally:
        h.close()


def test_re_export_preserves_identity_same_backing_no_map():
    # Frozen model §5/§8: canonical identity is invariant across every edge. Re-exporting an L4-owned
    # backing for forwarding keeps the SOURCE identity (owner_instance_id / path / buffer_id /
    # generation) and the same backing, only stripping the mapping.
    l4 = mint_owner_instance_id()
    src = create_host_shared_buffer(64, l4, buffer_id=7, owner_worker_path="L4")
    try:
        sdesc = src.to_descriptor()
        hp = re_export(sdesc)
        assert hp.identity == src.identity  # identity invariant across the edge
        assert hp.backend_kind == BackendKind.POSIX_SHM
        assert hp.body == sdesc.body and hp.nbytes == 64  # same backing
        assert hp.shm is None and hp.base == 0  # no map (lazy — a compute leaf maps)
        # a tensor built from H' carries the source identity + the same shm body, so L2 can materialize it
        r = hp.tensor(shapes=(16,), dtype=DataType.FLOAT32)
        assert r.buffer.identity == src.identity
        assert r.buffer.body == sdesc.body
    finally:
        src.close()


def test_device_malloc_wrap_materialize():
    # A device pointer (from orch.malloc) wrapped as DEVICE_MALLOC: materializes to the pointer with
    # no map, address_space DEVICE (-> a child_memory Tensor).
    oid = mint_owner_instance_id()
    h = wrap_device_malloc(0xDEAD0000, 4096, oid, buffer_id=3, owner_worker_path="L3")
    assert h.backend_kind == BackendKind.DEVICE_MALLOC
    assert h.address_space == AddressSpace.DEVICE
    assert h.shm is None and h.base == 0xDEAD0000
    reg = ImportRegistry()
    imp = reg.materialize(h.to_descriptor())
    assert imp.base == 0xDEAD0000
    assert imp.address_space == AddressSpace.DEVICE
    assert imp.shm is None


def test_materialize_remote_sidecar_rejected():
    desc = BufferDescriptor(
        identity=_identity(),
        address_space=AddressSpace.HOST,
        access=AccessMode.READWRITE,
        backend_kind=BackendKind.REMOTE_SIDECAR,
        nbytes=8,
    )
    reg = ImportRegistry()
    with pytest.raises(ValueError, match="REMOTE_SIDECAR"):
        reg.materialize(desc)


def test_owner_instance_ids_are_distinct():
    ids = {mint_owner_instance_id() for _ in range(64)}
    assert len(ids) == 64
    assert all(len(i) == OWNER_INSTANCE_ID_BYTES for i in ids)


@pytest.mark.parametrize(
    "space,backend",
    [
        (AddressSpace.HOST, BackendKind.VMM_WINDOW),
        (AddressSpace.HOST, BackendKind.DEVICE_MALLOC),
        (AddressSpace.DEVICE, BackendKind.FORK_SHM),
        (AddressSpace.DEVICE, BackendKind.POSIX_SHM),
    ],
)
def test_descriptor_rejects_bad_capability_combo(space, backend):
    # §4.1 capability matrix: an unsupported address_space×backend_kind fails at construction (before
    # dispatch, before it can ride the wire).
    with pytest.raises(ValueError, match="capability"):
        BufferDescriptor(
            identity=_identity(),
            address_space=space,
            access=AccessMode.READWRITE,
            backend_kind=backend,
            nbytes=64,
            body=b"",
        )


def test_descriptor_accepts_legal_combos():
    for space, backend in [
        (AddressSpace.HOST, BackendKind.FORK_SHM),
        (AddressSpace.HOST, BackendKind.POSIX_SHM),
        (AddressSpace.DEVICE, BackendKind.VMM_WINDOW),
        (AddressSpace.DEVICE, BackendKind.DEVICE_MALLOC),
        (AddressSpace.HOST, BackendKind.REMOTE_SIDECAR),
        (AddressSpace.DEVICE, BackendKind.REMOTE_SIDECAR),
    ]:
        BufferDescriptor(_identity(), space, AccessMode.READWRITE, backend, 64, b"")


# --- the shared validator, on the paths Python can reach --------------------------------------
#
# `validate_tensor` guards two boundaries: construction (here) and blob decode, which after the wire
# flip is `TaskArgsView::tensors(i)` in task_args.h. Both are C++, and Python has no way to turn
# bytes into a Tensor at all — so the malformed-bytes cases (bad magic, unknown backend tag,
# generation 0, body_len past the array, a view that does not fit) are exercised where they can be
# built: tests/ut/cpp/types/test_buffer.cpp. What is reachable from here is the construction gate.


def test_construction_rejects_a_view_past_the_backing():
    h = create_host_shared_buffer(64, mint_owner_instance_id(), buffer_id=1)
    try:
        h.tensor(shapes=(16,), dtype=DataType.FLOAT32)  # exactly 64 B: fits
        with pytest.raises(ValueError, match="past the backing"):
            h.tensor(shapes=(17,), dtype=DataType.FLOAT32)
        with pytest.raises(ValueError, match="past the backing"):
            h.tensor(shapes=(16,), dtype=DataType.FLOAT32, byte_offset=4)
    finally:
        h.close()


def test_construction_rejects_a_zero_stride():
    # strides are element strides and strictly > 0: broadcast and negative step are unsupported, and
    # a 0 would make two coordinates alias without the overlap map ever seeing it.
    h = create_host_shared_buffer(256, mint_owner_instance_id(), buffer_id=1)
    try:
        with pytest.raises(ValueError, match="stride"):
            h.tensor(shapes=(4, 8), dtype=DataType.FLOAT32, strides=(8, 0))
    finally:
        h.close()
